apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+    * @dev Close an action
+    * @param _actionId Identification number of the action being closed
+    * @param _action Action instance being closed
+    */
+    function _closeAction(uint256 _actionId, Action storage _action) internal {
+        _action.closed = true;
+        emit ActionClosed(_actionId);
+    }
+
+    /**
+    * @dev Challenge an action
+    * @param _actionId Identification number of the action being challenged
+    * @param _action Action instance being challenged
+    * @param _challenger Address challenging the action
+    * @param _requirement Collateral requirement instance applicable for the challenge
+    * @param _settlementOffer Amount of collateral tokens the challenger would accept for resolving the dispute without involving the arbitrator
+    * @param _finishedSubmittingEvidence Whether the challenger is finished submitting evidence with the challenge context
+    * @param _context Link to a human-readable text providing context for the challenge
+    * @return Identification number for the created challenge
+    */
+    function _createChallenge(
+        uint256 _actionId,
+        Action storage _action,
+        address _challenger,
+        CollateralRequirement storage _requirement,
+        uint256 _settlementOffer,
+        bool _finishedSubmittingEvidence,
+        bytes _context
+    )
+        internal
+        returns (uint256)
+    {
+        // Store challenge
+        uint256 challengeId = challengesLength++;
+        Challenge storage challenge = challenges[challengeId];
+        challenge.actionId = _actionId;
+        challenge.challenger = _challenger;
+        challenge.context = _context;
+        challenge.settlementOffer = _settlementOffer;
+        challenge.endDate = getTimestamp64().add(_requirement.challengeDuration);
+        challenge.challengerFinishedEvidence = _finishedSubmittingEvidence;
+
+        // Transfer challenge collateral
+        _transferFrom(_requirement.token, _challenger, _requirement.challengeAmount);
+
+        // Transfer half of the Arbitrator fees
+        (, , IArbitrator arbitrator) = _getSettingFor(_action);
+        (, ERC20 feeToken, uint256 feeAmount) = arbitrator.getDisputeFees();
+        uint256 arbitratorFees = feeAmount / 2;
+        challenge.arbitratorFeeToken = feeToken;
+        challenge.arbitratorFeeAmount = arbitratorFees;
+        _transferFrom(feeToken, _challenger, arbitratorFees);
+        return challengeId;
+    }
+
+    /**
+    * @dev Dispute an action
+    * @param _action Action instance to be disputed
+    * @param _challenge Current challenge instance for the action
+    * @return _arbitrator Address of the IArbitrator applicable for the action
+    * @return _metadata Metadata content to be used for the dispute
+    * @return Identification number of the dispute created in the arbitrator
+    */
+    function _createDispute(Action storage _action, Challenge storage _challenge, IArbitrator _arbitrator, bytes memory _metadata)
```

Comment:
Since this is doing a lot more than just creating the dispute (token transfers/etc), maybe we can inline this function into `disputeAction()`?

Reply:
Stack too deep issues my friend haha. I like the abstraction to be honest, would you prefer using a more white-box name?

Reply:
Nope, on a second look this is good!